python/ray/dag/compiled_dag_node.py

```diff
@@ -946,13 +968,55 @@ def _preprocess(self) -> None:
                     # Add all readers to the NCCL group.
                     nccl_actors.add(downstream_actor_handle)
 
-        # If there were type hints indicating transport via NCCL, initialize
-        # the NCCL group on the participating actors.
         nccl_actors = list(nccl_actors)
         if None in nccl_actors:
             raise ValueError("Driver cannot participate in the NCCL group.")
-        if nccl_actors and self._nccl_group_id is None:
+
+        # Initialize a NCCL group for each set of actors. A set of actors can be
+        # calling P2P send/recv or collective methods.
+        actors_to_nccl_group_id: Dict[FrozenSet["ray.actor.ActorHandle"], str] = {}
+
+        # If a custom NCCL group is specified for P2P actors, initialize and cache
+        # the NCCL group ID.
+        if nccl_actors and self._custom_nccl_group:
             self._nccl_group_id = _init_nccl_group(nccl_actors, self._custom_nccl_group)
+            actors = frozenset(nccl_actors)
+            actors_to_nccl_group_id[actors] = self._nccl_group_id
+
+        # If a custom NCCL group is specified for collective actors, initialize and
+        # cache the NCCL group ID.
+        for collective_group in nccl_collective_groups:
+            type_hint = collective_group.type_hint
+            if type_hint.get_custom_nccl_group():
+                nccl_group_id = collective_group.init_nccl_group()
+                actors = frozenset(collective_group.actor_handles)
+                if actors not in actors_to_nccl_group_id:
+                    actors_to_nccl_group_id[actors] = nccl_group_id
+
+        # If a NCCL group for P2P actors is not initialized, initialize and cache
+        # the NCCL group ID.
+        if nccl_actors and self._nccl_group_id is None:
```

Comment:
Is the reason for splitting this code from the above block in case a custom group passed to collective ops can be used as the default group for p2p?

If so, please make it clear in the comments that the order matters, i.e. all custom groups must be initialized before we decide what group to use for other ops.

Reply:
Yes. Updated comments.